### ISSUE TITLE
Support for flow decorators on methods

### ISSUE THREAD

**Issue by zanieb** [PrefectHQ/prefect#8517]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

Currently, when a method in a class is decorated with `@flow`, it is not bound as a method of the class and therefore does not receive `self` during calls.

### Describe the proposed behavior

The flow decorator should support use on class and instance methods. The user should not have to manually provide `cls` or `self`.

### Example Use

Users may define a class with namespaced configuration then define a flow as a method on the class which uses `self` to access the configuration.

### Additional context

Here's an example that covers some of the relevant behavior:

```python
import inspect


def decorator(fn):
    # "fn" is not yet bound to the class, it's just a function
    print("ismethod", inspect.ismethod(fn))
    # calling "fn" here would result in a missing 'self'
    return fn


class Test:
    @decorator
    def foo(self):
        print(self)


# Since foo is a function, it is registered as a method of the class on initialization
print("Test.foo", Test.foo)
print("ismethod", inspect.ismethod(Test().foo))

# Calling 'foo' here works fine
Test().foo()

print("-----")


from functools import partial
from prefect.flows import flow


class Test:
    def __init__(self) -> None:
        # On class initialization, we redeclare the bar attribute to be bound to this
        # instance so it receives `self` on calls
        self.bar = partial(self.bar, self)

    @flow
    def bar(self):
        # When the flow is defined, Prefect does not know it is in a class
        print(self)


# bar is a flow not a function, so it is not registered as a method of the class
print("Test.bar", Test.bar)
print("ismethod", inspect.ismethod(Test().bar))

# this works because we've bound the flow to the class
Test().bar()


print("-----")

# Alternatively, we can use a `get_flow` pattern

class Test:
    def get_bar(self):
        @flow
        def bar():
            print(self)

        return bar


# this works because we've created the flow within the scope of a bound method
bar = Test().get_bar()
bar()

```

**Comment by austinweisgrau** [comment]
I'm not sure if this example would be included in the spec above, but it would be nice if a classmethod could also be decorated as a task.

e.g.
```
from prefect import task

class SomeObject:

    @task
    @classmethod
    def constructor(cls, params) -> 'SomeObject':
        result = cls()
        ...
        return result
```

**Comment by jlowin** [comment]
For posterity's sake, this issue was closed in #13944 which added support for instance/class/staticmethods for both flows and tasks.
